### DISCUSSION TITLE
How do I keep indent of a node’s actual start and not of its line?

### DISCUSSION THREAD

**Discussion by gregorias** [Wansmer/treesj#163] [Q&A] [answered]
Would it be possible to add an option to change the default indentation start for splitting? Currently, TreeSJ uses the first non-whitespace character of the current line. It would enable my use-case to change that to the first character of the target node.

I intend to write a split-join config for Haskell, but I can’t make it work with TreeSJ indentation behaviour.

Consider a Haskell program like this:

```haskell
main = do
  let x = [1, 2, 3]
  putStrLn $ show x
```

with a Haskell config like this:

```lua
haskell = {
  list = {
    both = {
      separator = ",",
    },
  }
}
```

I get: 

```haskell
main = do
  let x = [
    1,
    2,
    3,
  ]
  putStrLn $ show x
```

This is incorrect, because in a let statement, following lines must be indented more than the binding’s name (`x`). Something like this:

```haskell
main = do
  let x = [
            1,
            2,
            3,
          ]
  putStrLn $ show x
```

**Comment by Wansmer** [comment]
You can add the necessary indent using the `format_tree` option. Also, check the presets and tests for `nix`, as far as  I know, Haskel and Nix have similar syntax.

**Reply by gregorias** [reply]
Thanks for the quick response. Sadly, I don’t see how both options help here.

Nix is not relevant, because it doesn’t suffer from indentation problem. [Nix uses semicolons to separate let-bindings, not indentation.](https://nix.dev/manual/nix/2.17/language/constructs)

AFAICT, `format_tree` doesn’t help us here, because it returns only the list node but TreeSJ’s indentation depends on the line where the list node starts. The relevant indentation information is not included in the inputs to `format_tree`. At least, the information is not there directly. 

I think that TreeSJ’s API could benefit from exposing the currently hardcoded logic for computing initial indentation, and turning that into a parameter. Do you think it’s a good approach and would you be open to a PR?

**Reply by Wansmer** [reply, marked as answer]
> Sadly, I don’t see how both options help here.

E.g:
```lua
        haskell = {
          list = {
            both = {
              separator = ",",
            },
            split = {
              format_tree = function(tsj)
                local len = tsj:child(1):range()[2] -- index of first character of first node
                for child in tsj:iter_children() do
                  if not (child:is_first() or child:is_omit()) then
                    local indent = not child:is_last() and (" "):rep(len + vim.fn.shiftwidth()) or (" "):rep(len)
                    child:update_text(indent .. child:text())
                  end
                end
              end,
            },
          },
        },
```


> would you be open to a PR

PR is always welcome as long as it's beneficial and doesn't break the existing user experience.
